### PR TITLE
fix: filter out empty arrays and objects from json

### PR DESCRIPTION
**Pull request by MeilakhSiarhei**
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12964)

### CODE REVIEW

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 1]
```diff
@@ -93,6 +93,19 @@ void ToJson(std::chrono::system_clock::time_point const& field,
   j[name] = std::to_string(m);
 }
 
+void removeEmptyArraysAndObjects(nlohmann::json& j) {
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:96]: The Google Style Guide requires this to be `RemoveEmptyArrayAndObjects` (note the initial uppercase).

Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:96]: I would have defined this function as:

```cc
nlohmann::json RemoveEmpty(nlohmann::json j) {
```

That way you can say:

```cc
return RemoveEmptyArraysAndObjects(nlohmann::json::parse( ... ));
```

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 2]
```diff
@@ -93,6 +93,19 @@ void ToJson(std::chrono::system_clock::time_point const& field,
...
+  for (auto item = j.begin(); item != j.end(); ) {
+    if ((item.value().is_array() || item.value().is_object()) && !item.value().empty()) {
+      removeEmptyArraysAndObjects(*item);
+    }
+    if (item.value().empty()) {
+      j.erase(item++);
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:102]: The documentation says:

```
Invalidates iterators and references at or after the point of the erase, including the end() iterator.
```

That suggests you cannot use `item` after calling `.erase(item++)`.

Did I miss something?

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 3]
```diff
@@ -106,15 +119,11 @@ nlohmann::json RemoveJsonKeysAndEmptyFields(
               [&parsed](std::string const& key) { return parsed == key; });
           return !discard;
         }
-        if (event == nlohmann::json::parse_event_t::object_end) {
-          return parsed != nullptr && !parsed.empty();
-        }
-        if (event == nlohmann::json::parse_event_t::array_end) {
-          return !parsed.empty();
-        }
```
Comment (jsrinnn) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: I am not sure I follow the intent behind replacing this with a recursive function. Can you explain?

Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: It seems that all you need is to capture a `counters = std::vector<int>{{0}}` value in the `remove_empty_call_back` (nit: should be `_callback`) and then do something like:

```cc
  if (event == nlohmann::json::parse_event_t::object_start || event == nlohmann::json::parse_event_t::array_start) {
    counters.push_back(0);
    return true;
  }
  if (event == nlohmann::json::parse_event_t::key) {
    auto keep = std::all_of(keys.begin(), keys.end(), [&parsed](auto const& k) { return parsed != key; });
    if (keep) ++counters.back();
    return keep;
  }
  if (event == nlohmann::json::parse_event_t::object_end || event == nlohman::json::parse_event_t::array_end) {
    auto count = counters.back();
    counters.pop_back();
    return count != 0;
  }
```

**Review by coryan — CHANGES_REQUESTED** [review]
I have some questions about whether this is using invalidated iterators and there may be a chance to simplify it too.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 3 above)
Comment (jsrinnn) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: Ah I see . looks like the empty objects part above may not be getting removed correctly.

The filter keys are removed in the rest_stub here : https://github.com/googleapis/google-cloud-cpp/blob/2cb3cdbf81a370601c122772db0a02c97b25daf5/google/cloud/bigquery/v2/minimal/internal/job_rest_stub.cc#L68 assuming they were set in the request.

Can you also add a check there to make sure the filtering is only called if the `json_filter_keys` in the request is= non-empty?

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 1 above)
Comment (MeilakhSiarhei) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc]: Done.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 2 above)
Comment (MeilakhSiarhei) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:102]: I found such example here
https://github.com/nlohmann/json/issues/884#issuecomment-352977847
Quite old, but I couldn't find anything better.
Also documentation says:
```
Return value
Iterator following the last removed element. If the iterator pos refers to the last element, the end() iterator is returned.
```
Also I added another value in tests after empty array/object. Works ok.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 3 above)
Comment (MeilakhSiarhei) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: It seems like changing the business logic. I'll update the behavior in the next PR with a unit test for it.

Comment (jsrinnn) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: I don't think changing like so should change the overall business logic. Filtering should only get applied if filtering keys have been set in the request.

SGTM for doing it in a follow-up PR.

```
std::string filtered_json = json_payload;
if (!request.json_filter_keys.empty()) {
filtered_json = RemoveJsonKeysAndEmptyFields(json_payload.dump(),
                                                    request.json_filter_keys());
}
```

Comment (MeilakhSiarhei) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: I tried such callback function:

```
if (event == nlohmann::json::parse_event_t::array_start) {
          return true; // false - I tried both options
        }
        if (event == nlohmann::json::parse_event_t::array_end) {
          return false;
        }
        return true;
```

works this way. Input: `{"query_parameters":[]}`, output `{"query_parameters":<discarded>}` - which is invalid json. So I don't see any way to remove empty arrays while parsing the string.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 4]
```diff
@@ -93,6 +93,20 @@ void ToJson(std::chrono::system_clock::time_point const& field,
   j[name] = std::to_string(m);
 }
 
+nlohmann::json RemoveEmptyArraysAndObjects(nlohmann::json j) {
+  for (auto item = j.begin(); item != j.end(); ) {
+    if ((item.value().is_array() || item.value().is_object()) && !item.value().empty()) {
+      item.value() = RemoveEmptyArraysAndObjects(*item);
```
Comment (coryan) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:99]: maybe use `std::move(*item)`, assume that is possible.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/bigquery/v2/minimal/internal/json_utils.cc:114]: Thanks for trying that out.

**Review by coryan — COMMENTED** [review]
LGTM, but check with @jsrinnn first.

**Review by MeilakhSiarhei — COMMENTED** [review]
I'll check with her after all presubmit checks/tests pass. Could you run it?

> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12964#-:-NhbIEq7-4Jr_kPjlOnL:b-c3kipb)* status: 0 of 2 files reviewed, 6 unresolved discussions (waiting on @coryan and @jsrinnn)

___
*[`google/cloud/bigquery/v2/minimal/internal/json_utils.cc` line 99 at r3](https://reviewable.io/reviews/googleapis/google-cloud-cpp/12964#gh-1371876326:-NhbISAs8U9BUG2OC04a:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/9092918fb80a53a17f639c88ebb40be385dac8ba/google/cloud/bigquery/v2/minimal/internal/json_utils.cc#L99)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

maybe use `std::move(*item)`, assume that is possible.  
</blockquote></details>

Done.

File: google/cloud/bigquery/v2/minimal/internal/json_utils.cc [hunk 5]
```diff
@@ -93,6 +93,20 @@ void ToJson(std::chrono::system_clock::time_point const& field,
   j[name] = std::to_string(m);
 }
 
+nlohmann::json RemoveEmptyArraysAndObjects(nlohmann::json j) {
+  for (auto item = j.begin(); item != j.end(); ) {
+    if ((item.value().is_array() || item.value().is_object()) && !item.value().empty()) {
+      item.value() = RemoveEmptyArraysAndObjects(std::move(*item));
```
Comment (devbww) [inline google/cloud/bigquery/v2/minimal/internal/json_utils.cc:99]: Tying the client call-stack depth to the nesting depth of user-supplied data sounds problematic.  Can we use a data-space stack instead?

**Comment by coryan** [comment]
You should probably run the `checkers` build locally to fix formatting, and rebase to pick the latest CI goodies.

**Comment by MeilakhSiarhei** [comment]
I close this PR as it fixes the issue only partially.
Full fix (making the library smart enough to populate json with only required values) is going to take too much time. It can be done in future.
